trio/_core/_thread_cache.py

```diff
@@ -1,7 +1,48 @@
 from threading import Thread, Lock
 import outcome
+import sys
+import ctypes
+import ctypes.util
 from itertools import count
 
+from typing import Callable, Optional, Tuple
+from functools import partial
+
+
+def get_os_thread_name_func() -> Optional[Callable[[Optional[int], str], None]]:
+    def namefunc(setname: Callable[[int, bytes], int], ident: Optional[int], name: str):
+        if ident is not None:
+            setname(ident, bytes(name[:15], "ascii", "replace"))
+
+    def darwin_namefunc(
+        setname: Callable[[bytes], int], ident: Optional[int], name: str
+    ):
+        if ident is not None:
+            setname(bytes(name[:15], "ascii", "replace"))
```

Comment:
You comment below that `on mac OSX pthread_setname_np does not take a thread id`, but the function does actually take that - bug?  If they *can* have the same function signatures then there's some duplicate code instead.

Reply:
The code calling namefunc will always pass a thread id, so it doesn't need to differ between platforms. But `setname` doesn't get `ident` passed to it.
I wasn't sure what to do if `ident is None` in `darwin_namefunc`. Documentation says that happens if the thread isn't started, but I don't really know what that means wrt naming the thread, but I went with the safe option of not doing anything in that case. (and idk if it *can* happen given Trio's usage of it)
Will add some comment to it.